lib/barbeque/message_queue.rb

```diff
@@ -12,20 +15,29 @@ def initialize(queue_name)
       @stop      = false
     end
 
-    # Receive a message and delete them all from SQS queue immediately.
-    # TODO: Stop removing them immediately to implement retry.
+    # Receive a message from SQS queue.
     # @return [Barbeque::Message::Base]
     def dequeue
-      while valid_messages.empty?
+      loop do
         return nil if @stop
-        messages = receive_messages
-        messages = reject_invalid_messages(messages)
-        valid_messages.concat(messages)
+        message = receive_message
+        if message
+          if message.valid?
+            return message
+          else
+            delete_message(message)
```

Comment:
Invalid messages are already logged and notified with ExceptionHandler.
https://github.com/cookpad/barbeque/blob/v0.7.0/lib/barbeque/message.rb